rpcs3/util/expr.hpp

```diff
@@ -0,0 +1,281 @@
+		{
+			// Variables i, j, k, l, m, n (six @ loop counters)
+			return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, ctrs[Cmd - U'i'], std::forward<S0>(s0), std::forward<SS>(stack)...);
+		}
+	}
+};
+
+// Loop start
+template <utf32_expr S, usz I>
+struct utf32_cmd<U'@', S, I>
+{
+	template <typename S0, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, SS&&... stack)
+	{
+		ctrs[0] = 0;
+
+		while (ctrs[0] != umax)
+		{
+			utf32_cmd<S.data[I], S, I + 1>::eval(ctrs + 1, std::forward<S0>(s0), std::forward<SS>(stack)...);
+
+			ctrs[0]++;
+		}
+
+		// TODO
+		return;
+	}
+};
+
+// Read current level loop counter
+template <utf32_expr S, usz I>
+struct utf32_cmd<U'o', S, I>
+{
+	template <typename S0, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, SS&&... stack)
+	{
+		return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, ctrs[-1], std::forward<S0>(s0), std::forward<SS>(stack)...);
+	}
+};
+
+// Loop unconditional terminator
+template <utf32_expr S, usz I>
+struct utf32_cmd<U';', S, I>
+{
+	template <typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], SS&&... stack)
+	{
+		ctrs[-1] = -1;
+		return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs - 1, std::forward<SS>(stack)...);
+	}
+};
+
+// Loop continue-else
+template <utf32_expr S, usz I>
+struct utf32_cmd<U'\\', S, I>
+{
+	template <typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(void) eval(u64 /*ctrs[]*/, SS&&... /*stack*/)
+	{
+		return;
+	}
+};
+
+// Loop "if-else" block
+template <utf32_expr S, usz I>
+struct utf32_cmd<U'?', S, I>
+{
+	template <char32_t Cmd, usz J, usz L = 0, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) skip(u64 ctrs[], SS&&... stack)
+	{
+		if constexpr (Cmd == U'\\' && L == 0)
+		{
+			return utf32_cmd<S.data[J], S, J + 1>::eval(ctrs, std::forward<SS>(stack)...);
+		}
+
+		if constexpr (Cmd == U'@')
+		{
+			return skip<S.data[J], J + 1, L + 1>(ctrs, std::forward<SS>(stack)...);
+		}
+
+		if constexpr (Cmd == U';')
+		{
+			if constexpr (L > 0)
+			{
+				return skip<S.data[J], J + 1, L - 1>(ctrs, std::forward<SS>(stack)...);
+			}
+			else
+			{
+				return;
+			}
+		}
+
+		skip<S.data[J], J + 1, L>(ctrs, std::forward<SS>(stack)...);
+	}
+
+	template <typename S0, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, SS&&... stack)
+	{
+		if (s0)
+		{
+			utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, std::forward<SS>(stack)...);
+		}
+		else
+		{
+			skip<S.data[I], S, I + 1>(std::forward<SS>(stack)...);
+		}
+	}
+};
+
+// Duplicate top stack value
+template <utf32_expr S, usz I>
+struct utf32_cmd<U':', S, I>
+{
+	template <typename S0, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, SS&&... stack)
+	{
+		return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, std::forward<S0>(s0), std::forward<S0>(s0), std::forward<SS>(stack)...);
```

Comment:
If the value passed is an rvalue, `std::forward<S0>(s0)` will be equivalent `std::move(s0)`, which would leave one of the two items empty for types that implement `std::move`
[demo](https://gcc.godbolt.org/z/WT6s9vzoE)

Assuming you weren't planning on allowing these exprs able to modify values, converting everything to `const T&` instead of `T&&` should fix the issue

Reply:
Casts don't move anything on their own, and I can't discriminate.